d2/src/test/java/com/linkedin/d2/balancer/dualread/UriPropertiesDualReadMonitorTest.java

```diff
@@ -114,6 +124,75 @@ public void testReportData() {
+                new ImmutablePair<>(URI_PROPERTIES_URI_1_AND_2, false),
+                new ImmutablePair<>(URI_PROPERTIES_URI_1_AND_2, true)
+            )
+        },
+        {
+          Arrays.asList(
+            new ImmutablePair<>(URI_PROPERTIES_1, false),
+            new ImmutablePair<>(URI_PROPERTIES_URI_1_AND_2, false),
+            new ImmutablePair<>(URI_PROPERTIES_1, true),
+            new ImmutablePair<>(URI_PROPERTIES_URI_1_AND_2, true)
+            )
+        },
+        {
+            Arrays.asList(
+                new ImmutablePair<>(URI_PROPERTIES_1, false),
+                new ImmutablePair<>(URI_PROPERTIES_1, true),
+                new ImmutablePair<>(URI_PROPERTIES_URI_1_AND_2, true),
+            new ImmutablePair<>(URI_PROPERTIES_URI_1_AND_2, false)
+            )
+        },
+        {
+            Arrays.asList(
+                new ImmutablePair<>(URI_PROPERTIES_1, false),
+                new ImmutablePair<>(URI_PROPERTIES_URI_1_AND_2, false),
+                new ImmutablePair<>(URI_PROPERTIES_1, true),
+                new ImmutablePair<>(URI_PROPERTIES_1, false),
+                new ImmutablePair<>(URI_PROPERTIES_URI_1_AND_2, true),
+                new ImmutablePair<>(URI_PROPERTIES_1, true)
+            )
+        }
+    };
+  }
+  @SuppressWarnings("ResultOfMethodCallIgnored")
+  @Test(dataProvider = "reportDataInMultiThreadsDataProvider")
+  public void testReportDataInMultiThreads(List<Pair<UriProperties, Boolean>> properties) {
+    UriPropertiesDualReadMonitorTestFixture fixture = new UriPropertiesDualReadMonitorTestFixture();
+    UriPropertiesDualReadMonitor monitor = fixture.getMonitor();
+
+    ExecutorService executor = Executors.newFixedThreadPool(2,
+        new NamedThreadFactory("UriPropertiesDualReadMonitorTest"));
+    properties.forEach(p -> executor.execute(() -> reportAndVerifyState(monitor, p.getLeft(), p.getRight())));
+    try {
+      executor.shutdown();
+      executor.awaitTermination(500, TimeUnit.MILLISECONDS);
+      // similarity eventually converge to 1
+      assertEquals((double) monitor.getMatchedUris() / (double) monitor.getTotalUris(), 1.0);
+    } catch (InterruptedException e) {
+      Assert.fail("Tasks were interrupted");
+    }
+  }
+
+  private void reportAndVerifyState(UriPropertiesDualReadMonitor monitor, UriProperties prop, boolean fromNewLb) {
+    monitor.reportData(CLUSTER_1, prop, fromNewLb);
+    // if reportData on the same cluster are NOT synchronized, the total uris and matched uris counts could become < 0.
+    // e.g: when total uris = 1, matched uris = 1, if reporting URI_PROPERTIES_URI_1_AND_2 for new and old Lbs are
+    // executed concurrently, total uris will decrement by 1 twice and become -1.
+    // We verify that doesn't happen no matter what order the data is reported between the old and new Lbs.
+    assertTrue(monitor.getTotalUris() >= 0);
+    assertTrue(monitor.getMatchedUris() >= 0);
```

Comment:
Can you figure out _exactly_ what these values should be and check that instead of just checking `0`?

Reply:
It can be deterministic as it depends on which next task is picked up, right?